Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectBuilder/MSBuildProjectBuilder.cs

```diff
@@ -57,20 +58,21 @@ public static Task<bool> BuildAllProjectsAsync(string arguments, IProgress<(int
         /// <param name="msBuildProjectReferences">The collection of MSBuild projects to build.</param>
         /// <param name="arguments">The additional arguments passed to MSBuild.</param>
         /// <returns>A task that will have a result of true if the build succeeds.</returns>
-        public static async Task<bool> BuildProjectsAsync(this IReadOnlyCollection<MSBuildProjectReference> msBuildProjectReferences, string arguments = MSBuildProjectBuilder.DefaultBuildArguments)
+        public static bool BuildProjects(this IReadOnlyCollection<MSBuildProjectReference> msBuildProjectReferences, string arguments = MSBuildProjectBuilder.DefaultBuildArguments)
         {
-            // Sequentialize all builds that show UI for progress.
-            await MSBuildProjectBuilder.buildLock.WaitAsync();
+            // This method blocks so it should only be possible to have one call at a time (unless it is called from the wrong thread).
+            Debug.Assert(!MSBuildProjectBuilder.isBuildingWithDefaultUI);
+            MSBuildProjectBuilder.isBuildingWithDefaultUI = true;
             try
             {
                 // When in batch mode, simply log the progress.
                 if (Application.isBatchMode)
                 {
-                    return await MSBuildProjectBuilder.BuildProjectsAsync(
+                    return MSBuildProjectBuilder.BuildProjectsAsync(
                         msBuildProjectReferences,
                         $"{arguments} /v:diagnostic",
                         new DelegateProgress<(int, (string progressMessage, ProgressMessageType progressMessageType) progressUpdate)>(report => MSBuildProjectBuilder.LogProgressMessage(report.progressUpdate.progressMessage, report.progressUpdate.progressMessageType)),
-                        CancellationToken.None);
+                        CancellationToken.None).GetAwaiter().GetResult();
```

Comment:
Is there a possibility that this will try to use the current thread for continuations? And thus result in a deadlock?